### DISCUSSION TITLE
Shorter condition statements

### DISCUSSION THREAD

**Discussion by GerbenVersluis** [dotnet/csharplang#1653] [Language Ideas] [answered]
We all know the following struggle when writing conditions:
 ```   csharp
public enum MyEnum
{
    typeA,
    typeB,
    typeC,
    typeD
}

var type = MyEnum.typeA;

if(type != MyEnum.typeB && type != MyEnum.typeC && type != MyEnum.typeD)
{
    // do stuff
}
```

Writing the `type !=` over and over again feels like duplicated code.
The following would be nicer:
_Explanation: condition is true when type IS NOT `MyEnum.typeB` AND `MyEnum.typeC` AND `MyEnum.typeC`_ 

``` csharp
if(type != (MyEnum.typeB && MyEnum.typeC && MyEnum.typeD))
{
    // do stuff
}
```

This would also work for other types besides enums like such:
_Explanation: condition is true when foo is equal to `some`, `random`, `value` or `bar`_ 
``` csharp
var foo = "bar";

if(foo == ("some" || "random" || "value" || "bar"))
{
     // do stuff
}
```
It should work for every operator like such:
_Explanation: condition is true when foo is greater than limitA AND limitB_ 
``` csharp
var limitA = 5;
var limitB = 10;
var foo = 15;

if(foo > (limitA && limitB))
{
   // do stuff
}
```

Would this be a welcome feature?
_ps: i'm not creative enough to find a fitting title for this 'issue' / 'request'_

**Comment by HaloFour** [comment]
This would break existing code as arbitrary types may already overload the `&&` and `||` operators.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/1350

```cs
var foo = "bar";

if (foo is "some" or "random" or "value" or "bar")
{
     // do stuff
}
```

**Comment by AustinBryan** [comment]
@ufcpp That is better because it obviously won't be breaking previous code as well as being more closer to English, less parens,  and even shorter.

**Comment by theunrepentantgeek** [comment]
There's greater precedent for syntax along these lines:

```
public enum MyEnum
{
   typeA,
   typeB,
   typeC,
   typeD
}

var type = MyEnum.typeA;

if(type in (MyEnum.typeB, MyEnum.typeC, MyEnum.typeD))
{
   // do stuff
}
```
*(Inspired by very faded memories of the Delphi enumeration and set syntax)*

This avoids the need to overload the meaning of `&&` and `||` in new and wonderful ways that are very likely to conflict with existing uses, wouldn't break existing code, and doesn't require a new keyword.

But, I'm not sure the language *needs* this - and I strongly suspect making it work with predictable performance would be challenging.

**Reply by tacosontitan** [reply]
I'm curious as to what makes you suspect that predictable performance for this would be challenging? I can sort-of see it with the potential need for various types of branch instructions but I'm curious on your thoughts there. I've thought C# needed an IN clause for quite some time now to be fair (and for more than just enumerated types) so I absolutely love this idea!

**Comment by jveselka** [comment]
I found Flags enums usefull in similar cases:
```C#
[Flags]
public enum MyEnum
{
   //meta values
   typeThatShouldNotDoStuff = 1,
   //real values
   typeA = 2,
   typeB = 4 | typeThatShouldNotDoStuff,
   typeC = 8 | typeThatShouldNotDoStuff,
   typeD = 16 | typeThatShouldNotDoStuff
}

var type = MyEnum.typeA;

if(!type.HasFlag(MyEnum.typeThatShouldNotDoStuff))
{
   // do stuff
}
```
It will look much nicer once you assign good semantic names to enum values.

**Comment by YairHalberstadt** [comment, marked as answer]
with `and`, `or`, `not`, `>` and `<` patterns, this is all possible in C# 9:

```csharp
if(type is not (MyEnum.typeB or MyEnum.typeC or MyEnum.typeD))
{
    // do stuff
}

if(foo is "some" or "random" or "value" or "bar")
{
     // do stuff
}

const int limitA = 5;
const int limitB = 10;
var foo = 15;

if(foo is > limitA and > limitB)
{
   // do stuff
}
```

However it is limited to only work on constants.

**Reply by tacosontitan** [reply]
> However it is limited to only work on constants.

That's a start though right?

**Comment by spydacarnage** [comment]
I use a `bool IsOneOf<T>(this T @this, params T[] values)` extension method for this (and still will, for when non-constant values are needed)
